### ISSUE TITLE
Missing commit in Seastar submodule

### ISSUE THREAD

**Issue by cmodk** [scylladb/scylladb#12719]
Hi

I am trying to run ScyllaDB on a Raspberry Pi 4, but it fails with a SIGILL. Thought i would debug a little and make a build myself, but i stranded quite fast. 

On e.g. branch next-5.1, when i do the 
`git submodule update --init --force --recursive`
it produces the following error:
`Fetched in submodule path 'seastar', but it did not contain 328edb2bce5d6a73c0f949942e07034b3eb59819. Direct fetching of that commit failed.`

If I checkout the tag scylla-5.1.0, the submodules checks out fine.

**Comment by avelanarius** [comment]
This particular Seastar commit (`328edb2bce5d6a73c0f949942e07034b3eb59819`) is indeed missing from `scylladb/seastar` repo (https://github.com/scylladb/seastar/commit/328edb2bce5d6a73c0f949942e07034b3eb59819), however it is present in `scylladb/scylla-seastar` (https://github.com/scylladb/scylla-seastar/commit/328edb2bce5d6a73c0f949942e07034b3eb59819).

You can modify `.gitmodules` file to point to `scylla-seastar` instead (by changing `url` field to `../scylla-seastar`):
https://github.com/scylladb/scylladb/blob/53366db6c66798808143b1e3170f2bb6dbcae5f6/.gitmodules#L1-L4
and then `git submodule update` should work correctly.

**Comment by cmodk** [comment]
That works, thanks

**Comment by michoecho** [comment]
> You can modify .gitmodules file to point to scylla-seastar instead (by changing url field to ../scylla-seastar):

@avelanarius You are confusing the data structures of submodules. There are three places where submodule URLs are specified: `.git/modules/$module_name/config`, `.git/config` and `.gitmodules`, in order from highest to lowest precedence.

- `.gitmodules` describes the upstream (default/intended) URLs of submodules. You shouldn't modify it — it should be correctly set by the upstream. And indeed, on `next-5.1` it is already set to `scylladb/scylla-seastar`:
https://github.com/scylladb/scylladb/blob/5e88421360d5e7bbde3ef8f7453a3879be801829/.gitmodules#L1-L3

- `.git/config` describes the list of "active" submodules — the ones you want to work with locally. `git submodule update` only touches modules described in `.git/config`.

- `.git/modules/$module_name` contains the actual git dir of the submodule. This is what all the actual git commands (e.g. fetch, checkout) see when the user or `git submodule` invoke them on the submodule.

`.gitmodules` is used to create a `submodule` entry in `.git/config` during `git submodule init` (or `git submodule update --init`. The `submodule` entry in `.git/config` is used by `git submodule update` to clone the submodule into `.git/modules/$module_name`.

 So the URLs and remotes actually used by `git submodule update` are not in `.gitmodules`, but in `.git/config` and `.git/modules/$module_name/config`. Since the user can use them to override upstream defaults, they are not automatically synchronized with `.gitmodules` — if your first invocation of `git submodule update --init` happens on `master`, the module will be registered in `.git/config` and cloned into `.git/modules` using the URL found in `master`'s `.gitmodules`, and it will stay that way until explicitly changed, even if you switch to a branch with a different `.gitmodules`.

But there is a command to sync the entries in `.git/config` and `.git/modules/$module_name/config` with `.gitmodules` automatically: `git submodule sync`.

**TL;DR**: modifying `.gitmodules` doesn't make sense here, as it's already set correctly.
When you switch branches, don't modify `.gitmodules`, but sync the local submodule configs with `.gitmodules` before the `update`. (E.g. by using `git submodules sync`).

**Comment by michoecho** [comment]
Hopefully the doubts have been answered. Closing.
